src/Language/PureScript/TypeChecker/Types.hs

```diff
@@ -462,6 +477,36 @@ infer' (TypedValue checkType val ty) = do
   ty' <- introduceSkolemScope <=< replaceAllTypeSynonyms <=< replaceTypeWildcards $ elabTy
   tv <- if checkType then withScopedTypeVars moduleName args (check val ty') else return (TypedValue' False val ty)
   return $ TypedValue' True (tvToExpr tv) ty'
+infer' (VisibleTypeApp val typeArg@(TypeWildcard _ _)) = do
+  TypedValue' _ val' valTy <- infer' val
+  (val'', valTy') <- instantiatePolyTypeWithUnknowns val' valTy
+  case unconsVtaTypeVar valTy' of
+    Just ((a, i, k, s, _), t) ->
+      pure $ TypedValue' True val'' $ ForAll a i k t s NotVtaTypeVar
+    Nothing ->
+      throwError . errorMessage $ CannotApplyExpressionOfTypeOnType valTy typeArg
+infer' (VisibleTypeApp val typeArg) = do
+  TypedValue' _ val' valTy <- infer' val
+  -- We instantiate early such that polykinded type variables:
+  --
+  -- forall (k :: Type) (@t :: k). Proxy t
+  --
+  -- are turned into:
+  --
+  -- forall (@t :: k?). Proxy t
+  --
+  -- This allows typeArg's kind to unify with the unknown k.
+  (val'', valTy') <- instantiatePolyTypeWithUnknowns val' valTy
+  -- `kindOf` here eliminates ParensInType in typeArg
```

Comment:
Is kindOf the best way to do this? It seems like a fair bit of work if all you want is to drop ParensInType. Why is it necessary? Kind-checking something only discard the kind doesn't sit right to me. Why discard the kind? Should we be using it elsewhere? Is there kind unification that needs to happen?

Reply:
Initially, there was kind unification in this routine, though I found that it was pretty inconsistent with certain cases (https://github.com/purescript/purescript/pull/4235#discussion_r815294743). If I remember correctly, `inferKind` which is called downstream, also elaborates the types it was passed so I'm making use of that in this routine as well.